### ISSUE TITLE
Loading .safetensors Lora

### ISSUE THREAD

**Issue by adhikjoshi** [huggingface/diffusers#3064]
### Describe the bug

I have downloaded lora from civitai which is in .safetensor format.

When i load it using below code,

pipe = StableDiffusionPipeline.from_pretrained("runwayml/stable-diffusion-v1-5", torch_dtype=torch.float16)
pipe.unet.load_attn_procs("lora.safetensors")

It throws error : KeyError: 'to_k_lora.down.weight'

  File "/workspace/server/tasks.py", line 346, in txt2img
    self.pipe.unet.load_attn_procs(embd, use_safetensors=True)
  File "/opt/conda/envs/ldm/lib/python3.8/site-packages/diffusers/loaders.py", line 224, in load_attn_procs
    rank = value_dict["to_k_lora.down.weight"].shape[0]
KeyError: 'to_k_lora.down.weight'

### Reproduction

pipe = StableDiffusionPipeline.from_pretrained("runwayml/stable-diffusion-v1-5", torch_dtype=torch.float16)
pipe.unet.load_attn_procs("lora.safetensors")

### System Info

Diffusers Version: 0.15.0.dev0

**Comment by alejobrainz** [comment]
Kohya-ss/sd-scripts has a nice mechanism for it, but it broke with 0.15, but you can for sure load A1111 LoRA Tensors with the function below on 0.14.0:

```
def apply_lora(pipe, lora_path, weight:float = 1.0):
    from safetensors.torch import load_file
    from sd-scripts.networks.lora import create_network_from_weights
    import torch
    
    vae = pipe.vae
    text_encoder = pipe.text_encoder
    unet = pipe.unet

    sd = load_file(lora_path)
    lora_network, sd = create_network_from_weights(weight, None, vae, text_encoder, unet, sd)
    lora_network.apply_to(text_encoder, unet)
    lora_network.load_state_dict(sd)
    lora_network.to("cuda", dtype=torch.float16)
```

but as of 0.15 it fails:

```
assert lora.lora_name not in names, f"duplicated lora name: {lora.lora_name}"
AssertionError: duplicated lora name: lora_unet_down_blocks_0_attentions_0_transformer_blocks_0_attn1_to_q

```

**Comment by adhikjoshi** [comment]
> [Kohya-ss/sd-scripts](https://github.com/kohya-ss/sd-scripts) has a nice mechanism for it, but it broke with 0.15, but you can for sure load A1111 LoRA Tensors with the function below on 0.14.0:
> 
> 
> 
> ```
> 
> def apply_lora(pipe, lora_path, weight:float = 1.0):
> 
>     from safetensors.torch import load_file
> 
>     from sd-scripts.networks.lora import create_network_from_weights
> 
>     import torch
> 
>     
> 
>     vae = pipe.vae
> 
>     text_encoder = pipe.text_encoder
> 
>     unet = pipe.unet
> 
> 
> 
>     sd = load_file(lora_path)
> 
>     lora_network, sd = create_network_from_weights(weight, None, vae, text_encoder, unet, sd)
> 
>     lora_network.apply_to(text_encoder, unet)
> 
>     lora_network.load_state_dict(sd)
> 
>     lora_network.to("cuda", dtype=torch.float16)
> 
> ```
> 
> 
> 
> but as of 0.15 it fails:
> 
> 
> 
> ```
> 
> assert lora.lora_name not in names, f"duplicated lora name: {lora.lora_name}"
> 
> AssertionError: duplicated lora name: lora_unet_down_blocks_0_attentions_0_transformer_blocks_0_attn1_to_q
> 
> 
> 
> ```

CC @haofanwang @sayakpaul

**Comment by sayakpaul** [comment]
Can someone provide LoRA file in the A1111 format? Providing as many relevant details associated to the file as possible would be great too.

**Comment by adhikjoshi** [comment]
> Can someone provide LoRA file in the A1111 format? Providing as many relevant details associated to the file as possible would be great too.

I have downloaded offset noise trained lora and uploaded its .safetensor on huggingface

https://huggingface.co/adhikjoshi/epi_noiseoffset

**Comment by sayakpaul** [comment]
Cc: @patrickvonplaten ^

**Comment by adhikjoshi** [comment]
Here is function i made from convert_lora_safetensor_to_diffusers.py to load lora on inference time.

```
import torch
from safetensors.torch import load_file

def load_lora_weights(pipeline, checkpoint_path):
    # load base model
    pipeline.to("cuda")
    LORA_PREFIX_UNET = "lora_unet"
    LORA_PREFIX_TEXT_ENCODER = "lora_te"
    alpha = 0.75
    # load LoRA weight from .safetensors
    state_dict = load_file(checkpoint_path, device="cuda")
    visited = []

    # directly update weight in diffusers model
    for key in state_dict:
        # it is suggested to print out the key, it usually will be something like below
        # "lora_te_text_model_encoder_layers_0_self_attn_k_proj.lora_down.weight"

        # as we have set the alpha beforehand, so just skip
        if ".alpha" in key or key in visited:
            continue

        if "text" in key:
            layer_infos = key.split(".")[0].split(LORA_PREFIX_TEXT_ENCODER + "_")[-1].split("_")
            curr_layer = pipeline.text_encoder
        else:
            layer_infos = key.split(".")[0].split(LORA_PREFIX_UNET + "_")[-1].split("_")
            curr_layer = pipeline.unet

        # find the target layer
        temp_name = layer_infos.pop(0)
        while len(layer_infos) > -1:
            try:
                curr_layer = curr_layer.__getattr__(temp_name)
                if len(layer_infos) > 0:
                    temp_name = layer_infos.pop(0)
                elif len(layer_infos) == 0:
                    break
            except Exception:
                if len(temp_name) > 0:
                    temp_name += "_" + layer_infos.pop(0)
                else:
                    temp_name = layer_infos.pop(0)

        pair_keys = []
        if "lora_down" in key:
            pair_keys.append(key.replace("lora_down", "lora_up"))
            pair_keys.append(key)
        else:
            pair_keys.append(key)
            pair_keys.append(key.replace("lora_up", "lora_down"))

        # update weight
        if len(state_dict[pair_keys[0]].shape) == 4:
            weight_up = state_dict[pair_keys[0]].squeeze(3).squeeze(2).to(torch.float32)
            weight_down = state_dict[pair_keys[1]].squeeze(3).squeeze(2).to(torch.float32)
            curr_layer.weight.data += alpha * torch.mm(weight_up, weight_down).unsqueeze(2).unsqueeze(3)
        else:
            weight_up = state_dict[pair_keys[0]].to(torch.float32)
            weight_down = state_dict[pair_keys[1]].to(torch.float32)
            curr_layer.weight.data += alpha * torch.mm(weight_up, weight_down)

        # update visited list
        for item in pair_keys:
            visited.append(item)

    return pipeline

```

Can use it like,


```
lora_model = lora_models + "/" + opt.lora + ".safetensors"
self.pipe = load_lora_weights(self.pipe, lora_model)
```

@sayakpaul @patrickvonplaten

**Comment by sayakpaul** [comment]
Thanks! Do you have the checkpoints with which we could test this?

**Comment by adhikjoshi** [comment]
> > Can someone provide LoRA file in the A1111 format? Providing as many relevant details associated to the file as possible would be great too.
> 
> 
> 
> I have downloaded offset noise trained lora and uploaded its .safetensor on huggingface
> 
> 
> 
> https://huggingface.co/adhikjoshi/epi_noiseoffset

This uploaded safetensor lora and others work well

**Comment by pdoane** [comment]
Thanks @adhikjoshi!  Getting a lot further with your function but the output is not matching what I would expect.  As a first guess, I would think this is the alpha handling as that is hard-coded to 0.75 but the LoRA's I'm using have .alpha keys in them.

**Comment by pdoane** [comment]
I updated the function from @adhikjoshi to use the .alpha elements and also added a multiplier that can be used to weight the LoRA overall.  Tested this on 4 random LoRAs I downloaded from CivitAI and it matches the output from Automatic1111:

```
def load_lora_weights(pipeline, checkpoint_path, multiplier, device, dtype):
    # load base model
    pipeline.to(device)
    LORA_PREFIX_UNET = "lora_unet"
    LORA_PREFIX_TEXT_ENCODER = "lora_te"
    # load LoRA weight from .safetensors
    state_dict = load_file(checkpoint_path, device=device)

    updates = defaultdict(dict)
    for key, value in state_dict.items():
        # it is suggested to print out the key, it usually will be something like below
        # "lora_te_text_model_encoder_layers_0_self_attn_k_proj.lora_down.weight"

        layer, elem = key.split('.', 1)
        updates[layer][elem] = value

    # directly update weight in diffusers model
    for layer, elems in updates.items():

        if "text" in layer:
            layer_infos = layer.split(LORA_PREFIX_TEXT_ENCODER + "_")[-1].split("_")
            curr_layer = pipeline.text_encoder
        else:
            layer_infos = layer.split(LORA_PREFIX_UNET + "_")[-1].split("_")
            curr_layer = pipeline.unet

        # find the target layer
        temp_name = layer_infos.pop(0)
        while len(layer_infos) > -1:
            try:
                curr_layer = curr_layer.__getattr__(temp_name)
                if len(layer_infos) > 0:
                    temp_name = layer_infos.pop(0)
                elif len(layer_infos) == 0:
                    break
            except Exception:
                if len(temp_name) > 0:
                    temp_name += "_" + layer_infos.pop(0)
                else:
                    temp_name = layer_infos.pop(0)

        # get elements for this layer
        weight_up = elems['lora_up.weight'].to(dtype)
        weight_down = elems['lora_down.weight'].to(dtype)
        alpha = elems['alpha']
        if alpha:
            alpha = alpha.item() / weight_up.shape[1]
        else:
            alpha = 1.0

        # update weight
        if len(weight_up.shape) == 4:
            curr_layer.weight.data += multiplier * alpha * torch.mm(weight_up.squeeze(3).squeeze(2), weight_down.squeeze(3).squeeze(2)).unsqueeze(2).unsqueeze(3)
        else:
            curr_layer.weight.data += multiplier * alpha * torch.mm(weight_up, weight_down)

    return pipeline
```

Example usage:

```
pipe = load_lora_weights(pipe, lora_path, 1.0, 'cuda', torch.float32)
```

**Comment by alejobrainz** [comment]
I tested on several custom-created Loras,  works great! Excellent work, @pdoane, thanks for sharing.

Quick question for the group.  is there a way to quickly unload a Lora weight from a loaded pipeline?  I want to maintain it in memory, and simply assign/remove Lora Embeddings on the fly after each inference.  Any pointers are appreciated. 

Thanks again!

Alejandro

**Comment by pdoane** [comment]
There are two options I can think of:

- Layer updating is a linear operation so it can be reversed by passing in a negative multiplier.  Because of floating-point rounding, there could be a gradual drift over time.

- You can make a copy of the tensor for each modified layer and restore it later.  As the LoRAs are small relative to the model, this is probably preferred (and I expect faster).

**Comment by alejobrainz** [comment]
I'll try approach #2

**Comment by alejobrainz** [comment]
Ugly, but worked for me.  Tested making 600 inferences switching between 12 Loras safetensors 50 times  on diffusers 0.15.1:

```
from safetensors.torch import load_file
from collections import defaultdict
from diffusers.loaders import LoraLoaderMixin
import torch 

current_pipeline = None
original_weights = {}

def load_lora_weights(pipeline, checkpoint_path, multiplier, device, dtype):
    global current_pipeline, original_weights
    
    if (pipeline != current_pipeline):
        backup = True
        current_pipeline = pipeline
        original_weights = {}    
    else:
        backup = False
    
    # load base model
    pipeline.to(device)
    LORA_PREFIX_UNET = "lora_unet"
    LORA_PREFIX_TEXT_ENCODER = "lora_te"
    # load LoRA weight from .safetensors
    state_dict = load_file(checkpoint_path, device=device)

    updates = defaultdict(dict)
    for key, value in state_dict.items():
        # it is suggested to print out the key, it usually will be something like below
        # "lora_te_text_model_encoder_layers_0_self_attn_k_proj.lora_down.weight"

        layer, elem = key.split('.', 1)
        updates[layer][elem] = value

    index = 0
    # directly update weight in diffusers model
    for layer, elems in updates.items():
        index += 1

        if "text" in layer:
            layer_infos = layer.split(LORA_PREFIX_TEXT_ENCODER + "_")[-1].split("_")
            curr_layer = pipeline.text_encoder
        else:
            layer_infos = layer.split(LORA_PREFIX_UNET + "_")[-1].split("_")
            curr_layer = pipeline.unet

        # find the target layer
        temp_name = layer_infos.pop(0)
        while len(layer_infos) > -1:
            try:
                curr_layer = curr_layer.__getattr__(temp_name)
                if len(layer_infos) > 0:
                    temp_name = layer_infos.pop(0)
                elif len(layer_infos) == 0:
                    break
            except Exception:
                if len(temp_name) > 0:
                    temp_name += "_" + layer_infos.pop(0)
                else:
                    temp_name = layer_infos.pop(0)

        # get elements for this layer
        weight_up = elems['lora_up.weight'].to(dtype)
        weight_down = elems['lora_down.weight'].to(dtype)
        alpha = elems['alpha']
        if alpha:
            alpha = alpha.item() / weight_up.shape[1]
        else:
            alpha = 1.0
        
        if (backup):
            original_weights[index] = curr_layer.weight.data.clone().detach()
        else:
            curr_layer.weight.data = original_weights[index].clone().detach()

        # update weight
        if len(weight_up.shape) == 4:
            curr_layer.weight.data += multiplier * alpha * torch.mm(weight_up.squeeze(3).squeeze(2), weight_down.squeeze(3).squeeze(2)).unsqueeze(2).unsqueeze(3)
        else:
            curr_layer.weight.data += multiplier * alpha * torch.mm(weight_up, weight_down)

    return pipeline

LoraLoaderMixin.load_lora_weights = load_lora_weights
```

**Comment by sayakpaul** [comment]
@pdoane thanks so much for your inputs and investigations! 

Do you mind sharing the `pipe` and `lora_path` you tested https://github.com/huggingface/diffusers/issues/3064#issuecomment-1512429695 with?

**Comment by pdoane** [comment]
@sayakpaul - followed up in e-mail.

**Comment by sayakpaul** [comment]
Thanks. However, I think having an end-to-end open example here would help the community a great deal to understand the nuances of the interoperability in a better manner.

**Comment by sayakpaul** [comment]
@pdoane come to think of it, would you be interested to improve our LoRA functionality to operate with the A1111 format as well? 

@patrickvonplaten recently incorporated similar support for our textual inversion scripts: https://github.com/huggingface/diffusers/tree/main/examples/textual_inversion

**Comment by alexblattner** [comment]
@alejobrainz how do you use your code for it to work with a prompt in the same way as A111?  I put this as prompt: "art by <lora:mngstle:1>" but it ignored it completely.

**Comment by alejobrainz** [comment]
For prompt weighting you can use [compel](https://github.com/damian0815/compel).  It's great and easy to use.  Just be sure to check out the syntax at https://github.com/damian0815/compel/blob/main/Reference.md

**Comment by alejobrainz** [comment]
Also, be mindful that the lora is embedded using the script.  you only need the keyword your Lora uses within the prompt.

**Comment by alexblattner** [comment]
@alejobrainz 
1. what would be the equivalent of (dog), (((cat)))?
2. is it possible to modify the word associated with the lora?
3. what would be the equivalent of just doing this "art by <lora:mngstle:1>"?

I am used to use automatic1111 and am developing a new interface using diffusers

**Comment by pdoane** [comment]
We're getting off topic quickly here, but Compel and A1111 are not equivalent when it comes to prompt weighting.  Roughly though, `(dog)` == `dog+` and `(((cat)))` == `cat+++`.  With Compel, parenthesis can be used for grouping but don't carry additional weighting information.

LoRAs modify the weights of the text encoder and unet, so they are not related to the prompt.  It's a bit of a weird design decision that some tools expose LoRA usage through the prompt at all.  Textual inversions do belong in the prompt, so that's likely where that comes from.

Depending on how far you want to go with your front-end, you are likely to run into some challenges.  Advanced usages like LoRAs, textual inversion, ControlNet, and the integration of all these features together is rapidly progressing but may not be ready out of the box.  What are the goals for your UI?  I may be working in a similar space.

**Comment by alexblattner** [comment]
@pdoane I want to create a full fledge AI comics maker with consistent characters, styles, etc... I knew there would be challenges which is why I'm doing it.

In A1111, when you train a Lora and have [name] as input text it uses the name of the lora and recognizes it. I am trying to do the same

**Comment by pdoane** [comment]
Compel can extract LoRAs referenced from prompts.  It'll be up to you to take that information and modify the pipeline though.

**Comment by alexblattner** [comment]
@pdoane is there any tutorial that could point me in the right direction on that?

**Comment by adhikjoshi** [comment]
> @pdoane is there any tutorial that could point me in the right direction on that?
> 
> EDIT: here's the full code to achieve what I tried to do:
> 
> ```
> import torch
> from safetensors.torch import load_file
> from collections import defaultdict
> def load_lora_weights(pipeline, checkpoint_path, multiplier, device, dtype):
>     LORA_PREFIX_UNET = "lora_unet"
>     LORA_PREFIX_TEXT_ENCODER = "lora_te"
>     # load LoRA weight from .safetensors
>     if isinstance(checkpoint_path, str):
> 
>         state_dict = load_file(checkpoint_path, device=device)
> 
>         updates = defaultdict(dict)
>         for key, value in state_dict.items():
>             # it is suggested to print out the key, it usually will be something like below
>             # "lora_te_text_model_encoder_layers_0_self_attn_k_proj.lora_down.weight"
> 
>             layer, elem = key.split('.', 1)
>             updates[layer][elem] = value
> 
>         # directly update weight in diffusers model
>         for layer, elems in updates.items():
> 
>             if "text" in layer:
>                 layer_infos = layer.split(LORA_PREFIX_TEXT_ENCODER + "_")[-1].split("_")
>                 curr_layer = pipeline.text_encoder
>             else:
>                 layer_infos = layer.split(LORA_PREFIX_UNET + "_")[-1].split("_")
>                 curr_layer = pipeline.unet
> 
>             # find the target layer
>             temp_name = layer_infos.pop(0)
>             while len(layer_infos) > -1:
>                 try:
>                     curr_layer = curr_layer.__getattr__(temp_name)
>                     if len(layer_infos) > 0:
>                         temp_name = layer_infos.pop(0)
>                     elif len(layer_infos) == 0:
>                         break
>                 except Exception:
>                     if len(temp_name) > 0:
>                         temp_name += "_" + layer_infos.pop(0)
>                     else:
>                         temp_name = layer_infos.pop(0)
> 
>             # get elements for this layer
>             weight_up = elems['lora_up.weight'].to(dtype)
>             weight_down = elems['lora_down.weight'].to(dtype)
>             alpha = elems['alpha']
>             if alpha:
>                 alpha = alpha.item() / weight_up.shape[1]
>             else:
>                 alpha = 1.0
> 
>             # update weight
>             if len(weight_up.shape) == 4:
>                 curr_layer.weight.data += multiplier * alpha * torch.mm(weight_up.squeeze(3).squeeze(2), weight_down.squeeze(3).squeeze(2)).unsqueeze(2).unsqueeze(3)
>             else:
>                 curr_layer.weight.data += multiplier * alpha * torch.mm(weight_up, weight_down)
>     else:
>         for ckptpath in checkpoint_path:
>             state_dict = load_file(ckptpath, device=device)
> 
>             updates = defaultdict(dict)
>             for key, value in state_dict.items():
>                 # it is suggested to print out the key, it usually will be something like below
>                 # "lora_te_text_model_encoder_layers_0_self_attn_k_proj.lora_down.weight"
> 
>                 layer, elem = key.split('.', 1)
>                 updates[layer][elem] = value
> 
>             # directly update weight in diffusers model
>             for layer, elems in updates.items():
> 
>                 if "text" in layer:
>                     layer_infos = layer.split(LORA_PREFIX_TEXT_ENCODER + "_")[-1].split("_")
>                     curr_layer = pipeline.text_encoder
>                 else:
>                     layer_infos = layer.split(LORA_PREFIX_UNET + "_")[-1].split("_")
>                     curr_layer = pipeline.unet
> 
>                 # find the target layer
>                 temp_name = layer_infos.pop(0)
>                 while len(layer_infos) > -1:
>                     try:
>                         curr_layer = curr_layer.__getattr__(temp_name)
>                         if len(layer_infos) > 0:
>                             temp_name = layer_infos.pop(0)
>                         elif len(layer_infos) == 0:
>                             break
>                     except Exception:
>                         if len(temp_name) > 0:
>                             temp_name += "_" + layer_infos.pop(0)
>                         else:
>                             temp_name = layer_infos.pop(0)
> 
>                 # get elements for this layer
>                 weight_up = elems['lora_up.weight'].to(dtype)
>                 weight_down = elems['lora_down.weight'].to(dtype)
>                 alpha = elems['alpha']
>                 if alpha:
>                     alpha = alpha.item() / weight_up.shape[1]
>                 else:
>                     alpha = 1.0
> 
>                 # update weight
>                 if len(weight_up.shape) == 4:
>                     curr_layer.weight.data += multiplier * alpha * torch.mm(weight_up.squeeze(3).squeeze(2), weight_down.squeeze(3).squeeze(2)).unsqueeze(2).unsqueeze(3)
>                 else:
>                     curr_layer.weight.data += multiplier * alpha * torch.mm(weight_up, weight_down)
>     return pipeline
> 
> import torch
> from diffusers import StableDiffusionControlNetPipeline, ControlNetModel, UniPCMultistepScheduler,ModelMixin
> from io import BytesIO
> from PIL import Image
> import torch.multiprocessing as mp
> import Loras
> controlnet = ControlNetModel.from_pretrained("lllyasviel/control_v11p_sd15_openpose", torch_dtype=torch.float16)
> pipe = StableDiffusionControlNetPipeline.from_pretrained(
>   "runwayml/stable-diffusion-v1-5", controlnet=controlnet, torch_dtype=torch.float16,safety_checker=None, requires_safety_checker=False,
> ).to("cuda")
> pipe=Loras.load_lora_weights(pipe, ['mngstle.safetensors','galgadot.safetensors'],1.0,'cuda',torch.float16)
> pipe.scheduler = UniPCMultistepScheduler.from_config(pipe.scheduler.config)
> 
> pipe.enable_xformers_memory_efficient_attention()
> pipe.enable_model_cpu_offload()
> buffer=open('gpose.png', 'rb')
> buffer.seek(0)
> image_bytes = buffer.read()
> images = Image.open(BytesIO(image_bytes))
> generator = torch.manual_seed(1)
> prompt="withLora(galgadot,1), manga, intricate, sharp focus, illustration, highly detailed, digital painting, concept art, matte, masterpiece, 8k, art by withLora(mngstle,1), black and white, monochrome"
> n_prompt="nsfw+, out of frame, multiple people, petite, loli, side view, profile, lowres, (bad anatomy, (bad hands)1.1)+, text, tattoo+, error, missing fingers, extra digit, fewer digits, cropped, worst quality, many people+++, low quality, normal quality, jpeg artifacts, signature, watermark, username, blurry, artist name,weird colors, (cartoon, 3d, bad art, poorly drawn, close up, (blurry)1.5)+, (disfigured, deformed, (extra limbs)1.5)+"
> fimage=pipe(
>     prompt,
>     images,
>     negative_prompt=n_prompt,
>     num_inference_steps=20,
>     generator=generator,
> )
> fimage = fimage.images[0]
> fimage.save('result.png', format='PNG')
> ```
> 
> this is controlnet with 2 loras used

How it would work with "(bad hands)1.1)+"  or any weighted prompts?

**Comment by alexblattner** [comment]
@adhikjoshi it's already weighted. a single + is the same as (word) and wordNUM is the same as (word:NUM)

**Comment by pdoane** [comment]
Diffusers does not apply prompt weighting to the prompt strings.  You need to use prompt embeds as explained here:

https://huggingface.co/docs/diffusers/main/en/using-diffusers/weighted_prompts

Remember that LoRAs adjust weights on the model and do not relate to position or positive/negative aspects of the prompt, so this is the line that matters:

```
pipe=Loras.load_lora_weights(pipe, ['mngstle.safetensors','galgadot.safetensors'],1.0,'cuda',torch.float16)
```

Compel has a prompt parser that can provide the list of LoRAs as well as the prompt embeds for diffusers.

**Comment by patrickvonplaten** [comment]
Would be super nice if we could add loading functionality to `load_lora_weights` for A1111 format. Fully agree with @sayakpaul statement above

**Comment by pdoane** [comment]
> We have a [load_lora_weights()](https://huggingface.co/docs/diffusers/main/en/api/loaders#diffusers.loaders.LoraLoaderMixin.load_lora_weights) utility which might be a better place.

Yes definitely!  I saw that after writing the message.  Maybe worth updating documentation to use it over the unet API?

> Do you mean passing in an actual state dict to the loading function? If so, it's something we actually do. See the docstrings of load_lora_weights() [here]

I think that might be something different.  I want to optimize changing the set of active LoRAs/weights.  As the LoRAs typically only adjust a subset of tensors in the model, it would be reasonable to restore a model back to its original state and then apply a new set.

Let me know If someone is working in the problem space here.  Otherwise I'll try to get some time for a PR in the next week or so.

**Comment by RustyKettle** [comment]
> I want to make sure I am using this correctly.
> 
> I would like to use Lykon/Dreamshaper for inpainting, so I need to load the safetensor weights at that location.
> 
> ```
> model = "Lykon/DreamShaper"
> pipeline = StableDiffusionInpaintPipeline.from_pretrained(
>     model,
>     torch_dtype=torch.float16,
> )
> pipeline = pipeline.load_lora_weights(model, use_safetensors=True, weight_name="DreamShaper_5_beta2_BakedVae-inpainting.inpainting.safetensors")
> pipeline = pipeline.to("cuda")
> ```
> 
> This gives the following error: `ValueError: None does not seem to be in the correct format expected by LoRA or Custom Diffusion training.`
> 
> My guess is that I'm calling it wrong.

I'm still trying to figure out how to use other inpainting weights, and am struggling. There doesn't seem to be any resources on this. This thread seems like the right place to discuss this, but I might be wrong. Should I start a new ticket?

**Comment by sayakpaul** [comment]
> Yes definitely! I saw that after writing the message. Maybe worth updating documentation to use it over the unet API?

For sure. This PR will hopefully fix it: https://github.com/huggingface/diffusers/pull/3180. 

> I think that might be something different. I want to optimize changing the set of active LoRAs/weights. As the LoRAs typically only adjust a subset of tensors in the model, it would be reasonable to restore a model back to its original state and then apply a new set.

I think we need to think a bit more about this design. I guess we have the following use-cases here. 

Let's consider we have two LoRAs (LoRA A and LoRA B), each being trained on concepts A and B, respectively. 

* Users could combine both and do some weighting to let the user control the effect. This is being discussed in: https://github.com/huggingface/diffusers/issues/2613. The way this is usually done is we just take the LoRAs and merge them sequentially to the affected main model blocks. Here, for example, after merging a single LoRA, we just repeat the process with the other LoRA(s):
https://github.com/huggingface/diffusers/blob/256e6960cbe8a6379ee396ca6317503a991b9bbe/scripts/convert_lora_safetensor_to_diffusers.py#L82

  Instead of taking `alpha`, I think we just apply the respective scalar LoRA weight coefficients. @pacman100 has worked on this in https://github.com/huggingface/peft. So, tagging him to see if he has any additional insights. 

* Users could first load, say, LoRA A, perform some generations, and then load LoRA B and repeat. 
  *  By default I think we should just clear the previously loaded LoRA to make room for the new LoRA being added. 
  * We provide APIs to the user to:
  > it would be reasonable to restore a model back to its original state and then apply a new set. 

> Let me know If someone is working in the problem space here. Otherwise I'll try to get some time for a PR in the next week or so.

Feel free to start the PR, we're more than happy to help :) 

Cc: @patrickvonplaten

**Comment by sayakpaul** [comment]
@RustyKettle, for https://github.com/huggingface/diffusers/issues/3064#issuecomment-1526885741, could you maybe open a new issue?

**Comment by showxu** [comment]
@sayakpaul  Hi, I created a PR https://github.com/huggingface/diffusers/pull/3294 to fix this issue. Tested with 3 .safetensor lora with diffusers pipeline.

**Comment by brrbrrry** [comment]
Hi, has anyone created anything to load multiple safetensors and apply them to a prompt? Otherwise is there any way to interact with the webui because the stable diffusion directory is broken on m1/m2 and I am only able to use a custom pytorch standalone webui atm. I would love to generate images with txt2img.py but I can't find it in the webui directory and wouldn't know how to load it so is there any way to interact with the already loaded webui with python scripts?

**Comment by patrickvonplaten** [comment]
Related: https://github.com/huggingface/diffusers/pull/3294

**Comment by takuma104** [comment]
I'm interested in this issue, and I've written own code for it. Since it uses hooks, it might not be possible to merge to Diffusers as-is, but this implementation supports multiple LoRAs and dynamic attachment/detachment. I'll think about whether it's possible to create a version without hooks as well.

https://gist.github.com/takuma104/e38d683d72b1e448b8d9b3835f7cfa44

**Comment by sayakpaul** [comment]
Seems like there are some entanglements. 

With #3294, I left my suggestions [here](https://github.com/huggingface/diffusers/pull/3294#issuecomment-1537798872) and [here](https://github.com/huggingface/diffusers/pull/3294#issuecomment-1537816327). 

Do we want to continue on top of #3294? But my concern is that it directly modifies the weights of the UNet which removes the flexibility part (see https://github.com/huggingface/diffusers/pull/3294#issuecomment-1537816327). 

So, need to discuss the best course of action here. 

@takuma104, if you want to happy to welcome a PR based on https://github.com/huggingface/diffusers/issues/3064#issuecomment-1538858776. 

@patrickvonplaten what are your thoughts?

**Comment by patrickvonplaten** [comment]
I don't think we should continue on top of #3294 - there are too many fundamental changes. Instead it'd be nice to open a new PR that cleanly allows loading A1111 weights I believe

**Comment by sayakpaul** [comment]
@takuma104 do you have a sample A111 LoRA weight file and a prompt to test with? I would like to prioritise working on the A1111 LoRA support in diffusers. For a first PR, may not have every possible case figured out but we will see.

**Comment by ghunkins** [comment]
I think transforming the `state_dict` in the `load_lora_weights` function to the `diffusers` format if `auto1111` format is found is the cleanest option here.

Looked into this, but stuck due to not being able to find documentation on the `diffusers` serialization format. `auto1111` seems pretty straightforward thankfully, but unsure how to convert to `diffusers`.

For diffusers format, I used @sayakpaul's testing repo `sayakpaul/dreambooth-text-encoder-test`.

```bash
!wget 'https://huggingface.co/sayakpaul/dreambooth-text-encoder-test/resolve/main/pytorch_lora_weights.bin'
```

```python
import torch
hf_lora = torch.load('pytorch_lora_weights.bin')
```

For auto1111 format, I used the most popular LoRA right now, MoXiN:

```bash
!wget 'https://civitai.com/api/download/models/14856' -O 'moxin.safetensors'
```

```python
import safetensors
sd_lora = safetensors.torch.load_file('moxin.safetensors')
```

Some intro code adapted from #3294 below to play with the serialization and transformation. **It is non-functional, as I don't fully understand the `diffusers` serialization format.**

```python
from typing import Dict

LORA_PREFIX_TEXT_ENCODER = 'lora_te'
LORA_PREFIX_UNET = 'lora_unet'

def convert_auto1111_to_diffusers_state_dict(state_dict: Dict):
    diffusers_state_dict = dict()
    for key in state_dict:
        # it is suggested to print out the key, it usually will be something like below
        # "lora_te_text_model_encoder_layers_0_self_attn_k_proj.lora_down.weight"

        # ignore alpha
        if ".alpha" in key:
            continue

        if "text" in key:
            layer_infos = key.split(".")[0].split(LORA_PREFIX_TEXT_ENCODER + "_")[-1].split("_")
            curr_layer = pipe.text_encoder
        else:
            layer_infos = key.split(".")[0].split(LORA_PREFIX_UNET + "_")[-1].split("_")
            curr_layer = pipe.unet

        # find the target layer
        transformed_name = []
        temp_name = layer_infos.pop(0)
        while len(layer_infos) > -1:
            try:
                curr_layer = curr_layer.__getattr__(temp_name)
                if len(layer_infos) > 0:
                    transformed_name.append(temp_name)
                    temp_name = layer_infos.pop(0)
                elif len(layer_infos) == 0:
                    transformed_name.append(temp_name)
                    break
            except Exception:
                if len(temp_name) > 0:
                    temp_name += "_" + layer_infos.pop(0)
                else:
                    transformed_name.append(temp_name)
                    temp_name = layer_infos.pop(0)

        # TODO: use the transformed name and key to
        # create the modified key
        transformed_key = '.'.join(transformed_name)

        # Needs more transformation, but I don't understand
        # the diffusers serialization well enough.
        # Example print:
        # up_blocks.3.attentions.1.transformer_blocks.0.attn1.to_k lora_down.weight
        print(transformed_key, '.'.join(key.split('.')[1:]))

        # ... once key is transformed, add to diffusers_state_dict
        diffusers_state_dict[transformed_key] = state_dict[key]

    return diffusers_state_dict
```

Another stickier consideration: the current monkey-patching of the `text_encoder` for LoRA doesn't allow for easy _removal_ of a LoRA to restore the previous model.

Happy to jump in more, but would need some documentation on the LoRA serialization format.

**Comment by sayakpaul** [comment]
> Looked into this, but stuck due to not being able to find documentation on the diffusers serialization format. auto1111 seems pretty straightforward thankfully, but unsure how to convert to diffusers.

Sorry for this. Maybe we should add a section [here](https://huggingface.co/docs/diffusers/main/en/training/lora) on the serialization format (cc: @patrickvonplaten). 

But trying to unblock you here. Previously, we only used to support LoRA fine-tuning of the UNet. You can find such a LoRA param file here: https://huggingface.co/patrickvonplaten/lora_dreambooth_dog_example/blob/main/pytorch_lora_weights.bin

There, we used to initialize the LoRA layers and just save them as regular torch/safetensors state dicts:

https://github.com/huggingface/diffusers/blob/ed616bd8a8740927770eebe017aedb6204c6105f/examples/dreambooth/train_dreambooth_lora.py#L647-#L666

(Note that the above code snippet is from an earlier [commit](https://github.com/huggingface/diffusers/commit/ed616bd8a8740927770eebe017aedb6204c6105f)) 

But now, the serialization format has changed with the introduction of [`save_lora_weights()`](https://huggingface.co/docs/diffusers/main/en/api/loaders#diffusers.loaders.LoraLoaderMixin.save_lora_weights). Now, to better distinguish between the UNet and the text params, we first initialize the LoRA layers like before. But during serialization, we follow:

https://github.com/huggingface/diffusers/blob/01c056f09441a8670d0a88f24e2d4fb4a2956ae8/src/diffusers/loaders.py#L1157-L1180

I hope this makes sense :-) 

Let me know if anything is unclear.

**Comment by takuma104** [comment]
@patrickvonplaten @sayakpaul Based on the insights gained from the code I wrote, I plan to open a PR using as much of the current Diffusers' implementation as possible. Before doing that, I've drafted a plan for the modification. Would it be okay to discuss it? It's a bit long, but it seems like this is about the amount of work involved.

1. Scope: Utilize as much of the current Diffusers' implementation as possible. 
Don't merge the weights, keep the LoRA weights as they are and compute them during the forward pass. This leaves room for future learning with Diffusers. Compatibility with A1111-like prompts (`<lora:xxx:1>`) is not included.

2. User API will use `pipe.load_lora_weights()`. 
This will automatically determine whether the format is A1111 and process accordingly.

3. Use `LoRAAttnProcessor` as is. 
There is not complete compatibility with the state_dict key values, but most of them can be processed with `LoRAAttnProcessor`. These should be able to be handled just by converting the key values. The code is still quite rough, but as an example with Unet, it would look like this: https://gist.github.com/takuma104/4d37c583e62b04dc250541bae6291f93

4. Alpha support for `LoRALinearLayer`. 
In the state_dict, there is an `.alpha` key corresponding to all `.lora_down`/`.lora_up` keys. This value is used as the scale of the output, calculated as `alpha/dim`. The `alpha` varies by file, sometimes the scale is `1.0`, other times it is not (e.g., `0.125`). The best implementation seems to be adding an `alpha` member to the `LoRALinearLayer` class.

5. Create new `LoRAbleLinear`, `LoRAbleConv2D` classes. 
In the state_dict, there are LoRA weights for `nn.Conv2D` and `nn.Linear` modules as well. These exist under keys like:
```
TextModel:
- text_model.encoder_layers.0.mlp_fc1
- text_model.encoder_layers.0.mlp_fc2

Unet:
- down_blocks.0.attentions.0.proj_in
- down_blocks.0.attentions.0.proj_out
- down_blocks.0.attentions.0.transformer_blocks_0.ff_net.0_proj
- down_blocks.0.attentions.0.transformer_blocks_0.ff_net.2
```
For these, I'm considering creating `LoRAbleLinear(nn.Linear)`, `LoRAbleConv2D(nn.Conv2D)` classes as a placeholder. When loading LoRA, create instances of these classes and dynamically replace them (text_encoder is a class of Transformers, so there doesn't seem to be any other way but monkey-patching), or replace them in advance on the Unet side. If you have any other ideas, please let me know.

Reference for Kohya's LoRA format info.:
LoRAs used in A1111 and similar are being distributed in places like CivitAI. These files are likely mostly produced by `[kohya-ss/sd-scripts](https://github.com/kohya-ss/sd-scripts)` trainer or its derivatives. Since Kohya-ss is a trainer derived from Diffusers 0.10.2, it uses this version of Diffusers internally, so the key values are quite similar. However, since it was created independently before Diffusers supported LoRA, there is not full compatibility with the current LoRA in Diffusers.

**Comment by pdoane** [comment]
@patrickvonplaten - I did a performance comparison of modifying weights vs a forward function.

My implementation for reference (roughly what is in this thread):

https://github.com/pdoane/simple-diffusion/blob/daed5e363f79e6e414ba990321f3e592c5b67b02/simple_diffusion/lora.py

And @takuma104's implementation previously linked that uses hooks, so I expect would have similar performance to attention processors:

https://gist.githubusercontent.com/takuma104/e38d683d72b1e448b8d9b3835f7cfa44/raw/22fd6b39b23d2049ba4ef283b2383f2d0750da4e/kohya_lora_loader.py

Testing a 512x512 image generation with 6 LoRAs:

- Modified weights: 2.15it/s
- Forward function: 1.33it/s

This is a big enough difference for me to justify direct weight modification.

**Comment by takuma104** [comment]
@sayakpaul I've been trying out generating about 100 samples with a high number of downloads using CivitAI's API. I won't list them all here, but as an example, one that particularly impressed me personally is a LoRA called [Light and Shadow](https://civitai.com/models/13239/light-and-shadow). I think it's probably a type of OffsetNoise. The code used to generate the pasted image is following:
https://gist.github.com/takuma104/82f267a9e0aa6d5581e57a241e5234a3

Without LoRA:
![light_and_shadow_not_apply](https://github.com/huggingface/diffusers/assets/10776/c4cb5f4e-69c0-45cb-b231-212d7c42e7be)

With Light and Shadow LoRA:
![light_and_shadow_apply](https://github.com/huggingface/diffusers/assets/10776/9dc450b7-ff52-4ca5-816b-97983618cc9a)

**Comment by takuma104** [comment]
@pdoane That's interesting. There's quite a difference in speed, about 61% of the original speed. In my environment (with xFormers, torch2.0) with my code, the speed dropped from 3.45it/s with no hooks to 2.64it/s with the LoRA hook. That was about 76% of the original speed.

**Comment by patrickvonplaten** [comment]
@pdoane could we leave this: https://github.com/huggingface/diffusers/issues/3064#issuecomment-1544165477 for another issue maybe to not entangle things too much here? :-) 

On a very global level:

- 1. We now have a LoRA serialization format of `diffusers` that **won't** change anymore
- 2. We have enabled LoRA for both text encoder and UNet
- 3. Now, the next step is to allow loading & converting A1111 LoRA serialization formats on the fly to the `diffusers` format.
- 4. Once this works I'm more than happy to look into speeding up LoRA for the forward pass (I also think that's important). 

So let's make this issue only about 3. and maybe open a new one for 4.? :-) 

Re: # 3.) 

@takuma104's plan here: https://github.com/huggingface/diffusers/issues/3064#issuecomment-1544159510 sounds like a great idea to me - @sayakpaul what do you think? 

I would maybe focus on those steps:

```
1. Scope:
Utilize as much of the current Diffusers' implementation as possible. Don't merge the weights, keep the LoRA weights as they are and compute them during the forward pass. This leaves room for future training with Diffusers. Compatibility with A1111-like prompts (`<lora:xxx:1>`) is not included.

3. User API will use `pipe.load_lora_weights()`. 
This will automatically determine whether the format is A1111 and process accordingly.

4. Use `LoRAAttnProcessor` as is. 
There is not complete compatibility with the state_dict key values, but most of them can be processed with `LoRAAttnProcessor`. These should be able to be handled just by converting the key values. The code is still quite rough, but as an example with Unet, it would look like this: https://gist.github.com/takuma104/4d37c583e62b04dc250541bae6291f93

5. Alpha support for `LoRALinearLayer`. 
In the state_dict, there is an `.alpha` key corresponding to all `.lora_down`/`.lora_up` keys. This value is used as the scale of the output, calculated as `alpha/dim`. The `alpha` varies by file, sometimes the scale is `1.0`, other times it is not (e.g., `0.125`). The best implementation seems to be adding an `alpha` member to the `LoRALinearLayer` class.
```

For now, they all sound great to me. 

The 6th step which involves creating LoRA for Conv layers is a bit out of scope for me at the moment here, so can we maybe try to only get A1111 LoRAs that only use Text Encoder Linear layers and UNet Linear layers working for now?

Some questions (maybe for @takuma104):
- 1.) What format do the most important A1111 LoRAs have. I assume not all of them have conv layer loras
- 2.) How many different formats are there in the A1111 world? 

Also cc @Chillee here as you were interesting in adding a LoRA layer to native PyTorch I think :-) 

@sayakpaul @takuma104 very interesting in hearing your thoughts here

**Comment by takuma104** [comment]
@alexblattner This code should be able to handle trigger words if the LoRA file being loaded contains LoRA weights for the TextEncoder.
https://gist.github.com/takuma104/e38d683d72b1e448b8d9b3835f7cfa44

**Comment by pdoane** [comment]
> could we leave this: https://github.com/huggingface/diffusers/issues/3064#issuecomment-1544165477 for another issue maybe to not entangle things too much here? :-)

Absolutely!  Getting functionality working is the right first step.  I was mostly curious about what the overhead was in case it was something that would impact design decisions.  I can also continue with a custom implementation in the near-term.

In terms of formats, on CivitAI there were 8 uploaded in Lycoris format last week and hundreds in the Kohya-ss LoRA format, so Lycoris support does not seem critical.

**Comment by alexblattner** [comment]
@takuma104 it seems like it does load the trigger words. not using a good checkpoint was the issue

**Comment by sanbuphy** [comment]
> Ugly, but worked for me. Tested making 600 inferences switching between 12 Loras safetensors 50 times on diffusers 0.15.1:
> 
> ```
> from safetensors.torch import load_file
> from collections import defaultdict
> from diffusers.loaders import LoraLoaderMixin
> import torch 
> 
> current_pipeline = None
> original_weights = {}
> 
> def load_lora_weights(pipeline, checkpoint_path, multiplier, device, dtype):
>     global current_pipeline, original_weights
>     
>     if (pipeline != current_pipeline):
>         backup = True
>         current_pipeline = pipeline
>         original_weights = {}    
>     else:
>         backup = False
>     
>     # load base model
>     pipeline.to(device)
>     LORA_PREFIX_UNET = "lora_unet"
>     LORA_PREFIX_TEXT_ENCODER = "lora_te"
>     # load LoRA weight from .safetensors
>     state_dict = load_file(checkpoint_path, device=device)
> 
>     updates = defaultdict(dict)
>     for key, value in state_dict.items():
>         # it is suggested to print out the key, it usually will be something like below
>         # "lora_te_text_model_encoder_layers_0_self_attn_k_proj.lora_down.weight"
> 
>         layer, elem = key.split('.', 1)
>         updates[layer][elem] = value
> 
>     index = 0
>     # directly update weight in diffusers model
>     for layer, elems in updates.items():
>         index += 1
> 
>         if "text" in layer:
>             layer_infos = layer.split(LORA_PREFIX_TEXT_ENCODER + "_")[-1].split("_")
>             curr_layer = pipeline.text_encoder
>         else:
>             layer_infos = layer.split(LORA_PREFIX_UNET + "_")[-1].split("_")
>             curr_layer = pipeline.unet
> 
>         # find the target layer
>         temp_name = layer_infos.pop(0)
>         while len(layer_infos) > -1:
>             try:
>                 curr_layer = curr_layer.__getattr__(temp_name)
>                 if len(layer_infos) > 0:
>                     temp_name = layer_infos.pop(0)
>                 elif len(layer_infos) == 0:
>                     break
>             except Exception:
>                 if len(temp_name) > 0:
>                     temp_name += "_" + layer_infos.pop(0)
>                 else:
>                     temp_name = layer_infos.pop(0)
> 
>         # get elements for this layer
>         weight_up = elems['lora_up.weight'].to(dtype)
>         weight_down = elems['lora_down.weight'].to(dtype)
>         alpha = elems['alpha']
>         if alpha:
>             alpha = alpha.item() / weight_up.shape[1]
>         else:
>             alpha = 1.0
>         
>         if (backup):
>             original_weights[index] = curr_layer.weight.data.clone().detach()
>         else:
>             curr_layer.weight.data = original_weights[index].clone().detach()
> 
>         # update weight
>         if len(weight_up.shape) == 4:
>             curr_layer.weight.data += multiplier * alpha * torch.mm(weight_up.squeeze(3).squeeze(2), weight_down.squeeze(3).squeeze(2)).unsqueeze(2).unsqueeze(3)
>         else:
>             curr_layer.weight.data += multiplier * alpha * torch.mm(weight_up, weight_down)
> 
>     return pipeline
> 
> LoraLoaderMixin.load_lora_weights = load_lora_weights
> ```

Thank you, this is very helpful to me. I guess the reason why Diffuser doesn't use this directly is that it directly modifies the weights and cannot achieve dynamic loading at runtime.

**Comment by sunhs** [comment]
@sanbuphy I think this already achieved dynamic loading at runtime, if I'm not mistaken. I'm using it in my service, where lora name can be specified per request.

**Comment by sanbuphy** [comment]
> @sanbuphy I think this already achieved dynamic (活跃) loading at runtime, if I'm not mistaken (弄错) . I'm using it in my service, where lora name can be specified per request.

 Hi ,How do you feel about the speed? I think it will take 4 to 5 seconds.
My question is that if I modify the weight in this way, doesn't it permanently change the weight of this pipeline? If I want to use the original one, I can't use it, can I?

**Comment by sunhs** [comment]
> > @sanbuphy I think this already achieved dynamic (活跃) loading at runtime, if I'm not mistaken (弄错) . I'm using it in my service, where lora name can be specified per request.
> 
>  Hi ,How do you feel about the speed? I think it will take 4 to 5 seconds.
> My question is that if I modify the weight in this way, doesn't it permanently change the weight of this pipeline? If I want to use the original one, I can't use it, can I?

You're right. You can't restore the original weight but it should be easy to achieve. The original code already restore the weight before applying the new lora weight.

**Comment by sanbuphy** [comment]
> > > @sanbuphy I think this already achieved dynamic (活跃) loading at runtime, if I'm not mistaken (弄错) . I'm using it in my service, where lora name can be specified per request.
> > 
> > 
> > Hi ,How do you feel about the speed? I think it will take 4 to 5 seconds.
> > My question is that if I 
> >           
> >             modify
> >             (修改)
> >           
> >            the weight in this way, doesn't it 
> >           
> >             permanently
> >             (永久)
> >           
> >            change the weight of this pipeline? If I want to use the original one, I can't use it, can I?
> 
> You're right. You can't restore (恢复) the original weight but it should be easy to achieve (实现) . The above code already restores the weight before applying the new lora weight.
> 
> As for speed, I haven't done any test but I didn't notice significant (显著) cost.

Hi! I found that it's just because I used the old version code below. I used an old version of this method. Let me try the new method. Thank you very much.

```python
def load_lora_weights(pipeline, checkpoint_path, multiplier, device, dtype):
    LORA_PREFIX_UNET = "lora_unet"
    LORA_PREFIX_TEXT_ENCODER = "lora_te"
    # load LoRA weight from .safetensors
    state_dict = load_file(checkpoint_path, device=device)

    updates = defaultdict(dict)
    for key, value in state_dict.items():
        # it is suggested to print out the key, it usually will be something like below
        # "lora_te_text_model_encoder_layers_0_self_attn_k_proj.lora_down.weight"

        layer, elem = key.split('.', 1)
        updates[layer][elem] = value

    # directly update weight in diffusers model
    for layer, elems in updates.items():

        if "text" in layer:
            layer_infos = layer.split(LORA_PREFIX_TEXT_ENCODER + "_")[-1].split("_")
            curr_layer = pipeline.text_encoder
        else:
            layer_infos = layer.split(LORA_PREFIX_UNET + "_")[-1].split("_")
            curr_layer = pipeline.unet

        # find the target layer
        temp_name = layer_infos.pop(0)
        while len(layer_infos) > -1:
            try:
                curr_layer = curr_layer.__getattr__(temp_name)
                if len(layer_infos) > 0:
                    temp_name = layer_infos.pop(0)
                elif len(layer_infos) == 0:
                    break
            except Exception:
                if len(temp_name) > 0:
                    temp_name += "_" + layer_infos.pop(0)
                else:
                    temp_name = layer_infos.pop(0)

        # get elements for this layer
        weight_up = elems['lora_up.weight'].to(dtype)
        weight_down = elems['lora_down.weight'].to(dtype)
        alpha = elems['alpha']
        if alpha:
            alpha = alpha.item() / weight_up.shape[1]
        else:
            alpha = 1.0

        # update weight
        if len(weight_up.shape) == 4:
            curr_layer.weight.data += multiplier * alpha * torch.mm(weight_up.squeeze(3).squeeze(2), weight_down.squeeze(3).squeeze(2)).unsqueeze(2).unsqueeze(3)
        else:
            curr_layer.weight.data += multiplier * alpha * torch.mm(weight_up, weight_down)

    return pipeline

```

**Comment by ashimatyuk** [comment]
@sayakpaul hi! Could you please tell me if HF has a tutorial for training lora for inpainting model and suitable for diffusers? If so, please provide link. thank you.

**Comment by auroxi** [comment]
Hi,

The above works fine for me, however if I call it a second time and pass a scale of 0, it doesn't revert back to the "no lora" version. How can I remove the applied lora after using the above? I need something like an unload_lora_weights function.

Thanks in advance for your help.

**Comment by patrickvonplaten** [comment]
I think we can close this one no @sayakpaul ?

**Comment by sayakpaul** [comment]
Yeah I think so. We have recently introduced support for loading A1111 formatted LoRAs (thanks to @takuma104?). 

We will continue to reiterate on top of it.

**Comment by notdanilo** [comment]
How can I update just the multiplier in `curr_layer.weight.data += multiplier * alpha * torch.mm(weight_up, weight_down)` without rebuilding the network?

**Comment by sayakpaul** [comment]
How would you like adjust the multipler?

**Comment by notdanilo** [comment]
> How would you like adjust the multiplier?

The same value is used on different layers, right? Say that I have initialized it with 1.0, then I just want to update it like `multiplier = 0.75`, updating everywhere it's referenced in the network.

Is it something possible to achieve?

**Comment by sayakpaul** [comment]
I think you would need to fetch the multiplier appropriately from the corresponding state dict accordingly for that. Since the multiplier only concerns the weight modification part, I don't think reinitialization would be required here.

**Comment by adhikjoshi** [comment]
With SDXL old lora won't be working.

Any workarounds?

**Comment by sayakpaul** [comment]
@sonaterai I think there are some extra keys and corresponding weight parameters for which we don't have support yet. See https://github.com/huggingface/diffusers/issues/3087. It also links to a couple of other similar threads. 

> With SDXL old lora won't be working.

It's not supposed as the corresponding UNet is different as far as I understand.

**Comment by sanbuphy** [comment]
> @sonaterai I think there are some extra keys and corresponding weight parameters for which we don't have support yet. See #3087. It also links to a couple of other similar threads.
> 
> > With SDXL old lora won't be working.
> 
> It's not supposed as the corresponding UNet is different as far as I understand.

If i train the Lora base on SDXL by myself, is that anything way to load the new Lora?

**Comment by sayakpaul** [comment]
@> If i train the Lora base on SDXL by myself, is that anything way to load the new Lora?

You should be able to. If not, please open a new thread with a reproducible set of instructions.

**Comment by kkwhale7** [comment]
> Here is function i made from convert_lora_safetensor_to_diffusers.py to load lora on inference time.
> 
> ```
> import torch
> from safetensors.torch import load_file
> 
> def load_lora_weights(pipeline, checkpoint_path):
>     # load base model
>     pipeline.to("cuda")
>     LORA_PREFIX_UNET = "lora_unet"
>     LORA_PREFIX_TEXT_ENCODER = "lora_te"
>     alpha = 0.75
>     # load LoRA weight from .safetensors
>     state_dict = load_file(checkpoint_path, device="cuda")
>     visited = []
> 
>     # directly update weight in diffusers model
>     for key in state_dict:
>         # it is suggested to print out the key, it usually will be something like below
>         # "lora_te_text_model_encoder_layers_0_self_attn_k_proj.lora_down.weight"
> 
>         # as we have set the alpha beforehand, so just skip
>         if ".alpha" in key or key in visited:
>             continue
> 
>         if "text" in key:
>             layer_infos = key.split(".")[0].split(LORA_PREFIX_TEXT_ENCODER + "_")[-1].split("_")
>             curr_layer = pipeline.text_encoder
>         else:
>             layer_infos = key.split(".")[0].split(LORA_PREFIX_UNET + "_")[-1].split("_")
>             curr_layer = pipeline.unet
> 
>         # find the target layer
>         temp_name = layer_infos.pop(0)
>         while len(layer_infos) > -1:
>             try:
>                 curr_layer = curr_layer.__getattr__(temp_name)
>                 if len(layer_infos) > 0:
>                     temp_name = layer_infos.pop(0)
>                 elif len(layer_infos) == 0:
>                     break
>             except Exception:
>                 if len(temp_name) > 0:
>                     temp_name += "_" + layer_infos.pop(0)
>                 else:
>                     temp_name = layer_infos.pop(0)
> 
>         pair_keys = []
>         if "lora_down" in key:
>             pair_keys.append(key.replace("lora_down", "lora_up"))
>             pair_keys.append(key)
>         else:
>             pair_keys.append(key)
>             pair_keys.append(key.replace("lora_up", "lora_down"))
> 
>         # update weight
>         if len(state_dict[pair_keys[0]].shape) == 4:
>             weight_up = state_dict[pair_keys[0]].squeeze(3).squeeze(2).to(torch.float32)
>             weight_down = state_dict[pair_keys[1]].squeeze(3).squeeze(2).to(torch.float32)
>             curr_layer.weight.data += alpha * torch.mm(weight_up, weight_down).unsqueeze(2).unsqueeze(3)
>         else:
>             weight_up = state_dict[pair_keys[0]].to(torch.float32)
>             weight_down = state_dict[pair_keys[1]].to(torch.float32)
>             curr_layer.weight.data += alpha * torch.mm(weight_up, weight_down)
> 
>         # update visited list
>         for item in pair_keys:
>             visited.append(item)
> 
>     return pipeline
> ```
> 
> Can use it like,
> 
> ```
> lora_model = lora_models + "/" + opt.lora + ".safetensors"
> self.pipe = load_lora_weights(self.pipe, lora_model)
> ```
> 
> @sayakpaul @patrickvonplaten

in my case in diffusers0.14.0, it gets wrong :
Traceback (most recent call last):
  File "demo_txt2img.py", line 88, in <module>
    demo.loadEngines(args.engine_dir, args.onnx_dir, args.onnx_opset,
  File "/workspace/demo/Diffusion/stable_diffusion_pipeline.py", line 288, in loadEngines
    model = obj.get_model()
  File "/workspace/demo/Diffusion/models.py", line 300, in get_model
    return load_lora_weights(basic_unet, self.lora, "unet")
  File "/workspace/demo/Diffusion/safetensors2bin.py", line 49, in load_lora_weights
    temp_name += "_" + layer_infos.pop(0)
IndexError: pop from empty list

can u help me ? thx

**Comment by sayakpaul** [comment]
Please use the `load_lora_weights()` API for this ([docs](https://huggingface.co/docs/diffusers/main/en/training/lora#supporting-different-lora-checkpoints-from-diffusers)). We don't maintain that script, so, sadly cannot offer any support.

**Comment by kkwhale7** [comment]
> Please use the `load_lora_weights()` API for this ([docs](https://huggingface.co/docs/diffusers/main/en/training/lora#supporting-different-lora-checkpoints-from-diffusers)). We don't maintain that script, so, sadly cannot offer any support.

however, i want to load lora for my unet and clip, not for the pipeline（Because I reused TensorRT's code）

**Comment by kkwhale7** [comment]
@sayakpaul

**Comment by SlZeroth** [comment]
I tried the load_lora_weights() but inference results are confuse. load_lora_weights load the text encoder params in the lora?

**Comment by kkwhale7** [comment]
> I tried the load_lora_weights() but inference results are confuse. load_lora_weights load the text encoder params in the lora?

yes , both clip and unet

**Comment by alirezaomneky** [comment]
I am having a similar issue. When I run pipe.load_lora_weights("lora_model_id"), where lora_model_id is the path to my model, I simply get this error:

KeyError: 'lora.down.weight'

Does anyone know how to fix this issue?

**Comment by sayakpaul** [comment]
Can you open a new issue with a reproducible code snippet?
